### PR TITLE
[docker] smaller image; fix file-editor app

### PR DESCRIPTION
**Pull request by lukeyeager**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,17 +1,23 @@
 FROM centos:7
 MAINTAINER Trey Dockendorf <tdockendorf@osc.edu>
-RUN yum install -y centos-release-scl lsof sudo sqlite-devel
-RUN yum install -y https://yum.osc.edu/ondemand/1.6/ondemand-release-web-1.6-1.el7.noarch.rpm
-RUN yum install -y ondemand
+RUN yum install -y \
+        centos-release-scl \
+        https://yum.osc.edu/ondemand/1.6/ondemand-release-web-1.6-1.el7.noarch.rpm \
+    && yum clean all
```
Comment (treydock) [inline Dockerfile]: Multiple `yum clean all` will cause the next `yum` command to take longer.  Only really need one `yum clean all` after all other `yum` commands.

File: Dockerfile [hunk 2]
```diff
@@ -1,17 +1,23 @@
...
+RUN yum install -y \
+        file \
+        httpd24-mod_ldap \
+        httpd24-mod_ssl \
+        lsof \
+        ondemand \
+        sqlite-devel \
+        sudo \
+    && yum clean all
```
Comment (treydock) [inline Dockerfile]: Multiple `yum clean all` will cause the next `yum` command to take longer.  Only really need one `yum clean all` after all other `yum` commands.

File: Dockerfile [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
+
```
Comment (treydock) [inline Dockerfile]: I'd move `yum clean all` to its own `RUN`.

**Comment by lukeyeager** [comment]
Hi @treydock, thanks for your review!

However, I strongly disagree with your suggestions. It is more important to have a small image (faster downloads for all users) than it is to have a quick docker build (you guys do it once and push it to docker hub).

For further reading on dockerfile best practices:

* https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run
* https://medium.com/@vaceletm/docker-layers-cost-b28cb13cb627
* https://gitlab.com/nvidia/cuda/blob/ubuntu18.04/10.1/base/Dockerfile

**Comment by treydock** [comment]
I agree but these two things save the same amount of space:

```
yum install package
yum clean all
yum install package2
yum clean all
```

`
```
yum install package
yum install package2
yum clean all
```

No space is saved if you execute `yum clean all` multiple times compared to executing it just once after all `yum` commands.  The command `yum clean all` clears out cached packages and repo metadata so a single execution can be done at the end of all other `yum` commands to save the space.  You can also delete `/var/cache/yum` to save space as sometimes `yum clean all` won't delete everything from that location.

**Comment by treydock** [comment]
The benefit to a single `yum clean all` is the previous `yum` commands can make use of the local cache before the `yum clean all` deletes the cache. So my suggested changes would save same space and keep the builds fast. I'd also recommend `rm -rf /var/cache/yum/*` if you want to save a little more extra space.

**Comment by treydock** [comment]
If you're seeing bigger images with a single `yum clean all` then I'd be very interested to know that as that may be something specific to Docker and each layer being smaller.  In that case multiple `yum clean all` may be necessary to keep each layer as small as possible.

**Comment by lukeyeager** [comment]
> No space is saved if you execute yum clean all multiple times compared to executing it just once after all yum commands. The command yum clean all clears out cached packages and repo metadata so a single execution can be done at the end of all other yum commands to save the space.

Removing the files in a second RUN cmd doesn't actually make the image smaller. The diff including those files is still included in one of the docker layers.

```console
$ cat Dockerfile.dd 
FROM ubuntu
# create big file; delete it in a second RUN cmd; adds 1GB to image size
RUN dd if=/dev/zero of=/zeros bs=1024 count=1000000
RUN rm /zeros
# create big file; delete it in the same RUN cmd; adds 0 to image size
RUN dd if=/dev/zero of=/zeros bs=1024 count=1000000 \
    && rm /zeros

$ docker build -t tmp -f Dockerfile.dd .

$ docker image ls tmp:latest
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
tmp                 latest              3e96c69b1261        6 minutes ago       1.13GB

$ docker history tmp:latest
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
3e96c69b1261        7 minutes ago       /bin/sh -c dd if=/dev/zero of=/zeros bs=1024…   0B                  
601a1ac052ff        8 minutes ago       /bin/sh -c rm /zeros                            0B                  
679034725d30        8 minutes ago       /bin/sh -c dd if=/dev/zero of=/zeros bs=1024…   1.02GB              
d131e0fa2585        2 months ago        /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B                  
<missing>           2 months ago        /bin/sh -c mkdir -p /run/systemd && echo 'do…   7B                  
<missing>           2 months ago        /bin/sh -c rm -rf /var/lib/apt/lists/*          0B                  
<missing>           2 months ago        /bin/sh -c set -xe   && echo '#!/bin/sh' > /…   745B                
<missing>           2 months ago        /bin/sh -c #(nop) ADD file:7ce84f13f11609a50…   102MB
```

> You can also delete /var/cache/yum to save space as sometimes yum clean all won't delete everything from that location.

You're right about this one. I had simply trusted other things I had read but didn't investigate myself. I'll update the PR.

**Comment by treydock** [comment]
I pushed `latest` and `1.6.1` to `ohiosupercomputer/ood` on Docker Hub with these changes.

**Pull request merged**